### ISSUE TITLE
Starting sandboxed systemd service inside an LXC container fails 

### ISSUE THREAD

**Issue by frederictobiasc** [lxc/lxc#4052]
# Required information

 * Distribution: Arch Linux
 * The output of
   * `lxc-start --version`: 4.0.11
   * `lxc-checkconfig`: 
```
LXC version 4.0.11
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Warning: newuidmap is not setuid-root
Warning: newgidmap is not setuid-root
Network namespace: enabled

--- Control groups ---
Cgroups: enabled
Cgroup namespace: enabled

Cgroup v1 mount points: 


Cgroup v2 mount points: 
/sys/fs/cgroup

Cgroup v1 systemd controller: missing
Cgroup v1 freezer controller: missing
Cgroup ns_cgroup: required
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, loaded
Advanced netfilter: enabled, loaded
CONFIG_NF_NAT_IPV4: missing
CONFIG_NF_NAT_IPV6: missing
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, not loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, not loaded
FUSE (for use with lxcfs): enabled, loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: 

Note : Before booting a new kernel, you can check its configuration
usage : CONFIG=/path/to/config /usr/bin/lxc-checkconfig
```
   * `uname -a`: Linux nb-istobic 5.15.12-arch1-1 #1 SMP PREEMPT Wed, 29 Dec 2021 12:04:56 +0000 x86_64 GNU/Linux
   * `cat /proc/self/cgroup`: 0::/user.slice/user-1000.slice/user@1000.service/app.slice/app-gnome-kitty-120347.scope
   * `cat /proc/1/mounts`
```
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
sys /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
dev /dev devtmpfs rw,nosuid,relatime,size=8126272k,nr_inodes=2031568,mode=755,inode64 0 0
run /run tmpfs rw,nosuid,nodev,relatime,mode=755,inode64 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
/dev/nvme1n1p3 / ext4 rw,relatime 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev,inode64 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
none /sys/fs/bpf bpf rw,nosuid,nodev,noexec,relatime,mode=700 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=30,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=11707 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime,pagesize=2M 0 0
mqueue /dev/mqueue mqueue rw,nosuid,nodev,noexec,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,nosuid,nodev,noexec,relatime 0 0
tracefs /sys/kernel/tracing tracefs rw,nosuid,nodev,noexec,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,nosuid,nodev,noexec,relatime 0 0
configfs /sys/kernel/config configfs rw,nosuid,nodev,noexec,relatime 0 0
systemd-1 /efi autofs rw,relatime,fd=47,pgrp=1,timeout=120,minproto=5,maxproto=5,direct,pipe_ino=10468 0 0
/dev/loop0 /var/lib/snapd/snap/bare/5 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop1 /var/lib/snapd/snap/core/11798 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop2 /var/lib/snapd/snap/core/11993 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop3 /var/lib/snapd/snap/core18/2246 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop4 /var/lib/snapd/snap/core18/2253 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop5 /var/lib/snapd/snap/gnome-3-28-1804/145 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop6 /var/lib/snapd/snap/gnome-3-28-1804/161 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop7 /var/lib/snapd/snap/gtk-common-themes/1515 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop8 /var/lib/snapd/snap/gtk-common-themes/1519 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop9 /var/lib/snapd/snap/skype/194 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop10 /var/lib/snapd/snap/skype/197 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop11 /var/lib/snapd/snap/wine-platform-3-stable/11 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop12 /var/lib/snapd/snap/wine-platform-3-stable/14 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop13 /var/lib/snapd/snap/wine-platform-runtime/277 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/loop14 /var/lib/snapd/snap/wine-platform-runtime/278 squashfs ro,nodev,relatime,errors=continue 0 0
/dev/nvme0n1p1 /mnt/c2d777dc-dbf7-4eeb-92be-1589a5c3abca ext4 rw,nosuid,nodev,relatime 0 0
tmpfs /tmp tmpfs rw,nosuid,nodev,size=8136764k,nr_inodes=1048576,inode64 0 0
/dev/nvme1n1p2 /boot vfat rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=ascii,shortname=mixed,utf8,errors=remount-ro 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,nosuid,nodev,noexec,relatime 0 0
tracefs /sys/kernel/debug/tracing tracefs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=1627352k,nr_inodes=406838,mode=700,uid=1000,gid=1000,inode64 0 0
gvfsd-fuse /run/user/1000/gvfs fuse.gvfsd-fuse rw,nosuid,nodev,relatime,user_id=1000,group_id=1000 0 0
jetbrains-toolbox /tmp/.mount_jetbra17dYxO fuse.jetbrains-toolbox ro,nosuid,nodev,relatime,user_id=1000,group_id=1000 0 0
portal /run/user/1000/doc fuse.portal rw,nosuid,nodev,relatime,user_id=1000,group_id=1000 0 0
run /run/snapd/ns tmpfs rw,nosuid,nodev,relatime,mode=755,inode64 0 0
nsfs /run/snapd/ns/skype.mnt nsfs rw 0 0
```

# Issue description
A sandboxed systemd service fails to start. This happens both in privileged and unprivileged containers. Log output:
```
Jan 06 11:47:06 playtime systemd[1]: Starting Synapse Matrix homeserver (master)...
Jan 06 11:47:06 playtime systemd[159]: synapse.service: Failed to set up mount namespacing: /run/systemd/unit-root/proc/sys/kernel/domainname: No such file or directory
Jan 06 11:47:06 playtime systemd[159]: synapse.service: Failed at step NAMESPACE spawning /usr/bin/python3: No such file or directory
Jan 06 11:47:06 playtime systemd[1]: synapse.service: Main process exited, code=exited, status=226/NAMESPACE
Jan 06 11:47:06 playtime systemd[1]: synapse.service: Failed with result 'exit-code'.
Jan 06 11:47:06 playtime systemd[1]: Failed to start Synapse Matrix homeserver (master).
```
When removing the sandboxing configuration ([override-hardened.conf](https://github.com/lxc/lxc/files/7821667/override-hardened.txt) stored in `/usr/lib/systemd/system/synapse.service.d/override-hardened.conf`), the unit starts without problems. 
Research showed that this problem occurred before (e.g. https://github.com/systemd/systemd/issues/17866), but the proposed fix (adding `security.nested=true` to the container config (`/var/lib/lxc/playtime/config`)) did not work for me.
I would be happy if you could give me some hints how to solve/debug this.

# Steps to reproduce

 1. Setup lxc according to the archwiki [documentation](https://wiki.archlinux.org/title/Linux_Containers).
 1. Create new container: `lxc-create -n playtime -t download -- --dist archlinux --release current --arch amd64`
 2. Start and attach
 3. Install matrix-synapse `pacman -Sy && pacman -S matrix-synapse`
 4. Create initial configuration:
 ```
  $ cd /var/lib/synapse
  $ sudo -u synapse python -m synapse.app.homeserver \
      --server-name my.domain.name \


      --config-path /etc/synapse/homeserver.yaml \
      --generate-config \
      --report-stats=yes
 ```
5. `chown -R synapse:synapse /etc/synapse`
6. `systemctl start synapse.service`

[Video](https://asciinema.org/a/uGACOpY6RTpNNoMiIxcs3rWFd)

# Information to attach

 - [X] any relevant kernel output ([dmesg.log](https://github.com/lxc/lxc/files/7821655/dmesg.log))
 - [X] container log ([lxclog.log](https://github.com/lxc/lxc/files/7821625/lxclog.log))
 - [X] the containers configuration file 
```                                                                                  
# Template used to create this container: /usr/share/lxc/templates/lxc-download
# Parameters passed to the template: --dist archlinux --release current --arch amd64
# Template script checksum (SHA-1): 985e49018f4037a7854e89877ba09c4fe6c57af5
# For additional config options, please look at lxc.container.conf(5)

# Uncomment the following line to support nesting containers:
lxc.include = /usr/share/lxc/config/nesting.conf
# (Be aware this has security implications)


# Distribution configuration
lxc.include = /usr/share/lxc/config/common.conf
lxc.arch = x86_64

# Container specific configuration
lxc.rootfs.path = dir:/var/lib/lxc/playtime/rootfs
lxc.uts.name = playtime

# Network configuration
lxc.net.0.type = veth
lxc.net.0.link = lxcbr0
lxc.net.0.flags = up
lxc.net.0.hwaddr = 00:16:3e:61:84:d2
```

**Comment by brauner** [comment]
I think that the AppArmor profile is stopping systemd from changing the mount propagation in a new mount namespace (//Cc @stgraber). If I turn of AppArmor things work but I see another failure:
```
Jan 06 12:39:07 playtime systemd[109]: synapse.service: Failed to set up mount namespacing: /run/systemd/unit-root/proc/sys/kernel/domainname: No such file or directory
Jan 06 12:39:07 playtime systemd[109]: synapse.service: Failed at step NAMESPACE spawning /usr/bin/python3: No such file or directory
```
which I don't know what this is about rn.

**Comment by frederictobiasc** [comment]
Isn't it the same failure I mentioned? I don't have AppArmor set up on the host.

**Comment by brauner** [comment]
Pasting what I commented in https://discuss.linuxcontainers.org/t/hardened-service-fails-to-start/13118/10?u=brauner here too:
Ok, I figured this out.

So in your `override-hardened.conf` you have the line
```
ProcSubset=pid
```
set. This line will cause `procfs` to be mounted with the `subset=pid` option. The `subset=pid` option means that only `/proc/<pid>` directories will be visible and no other `procfs` files.

Consequently, the `/run/systemd/unit-root/proc/sys/kernel/domainname` file will not be available to the `synapse.service`. So in order to have a functioning service you will need to disable `ProcSubset=pid` in your `override-hardened.conf`. The `subset=pid` mount option for `procfs` will be available on reletaviely new kernels. Older kernels that do not support this option will work.
